### PR TITLE
Provide PEM implementations for AsymmetricAlgorithm.

### PR DESCRIPTION
**Pull request by vcsjones**
Now that the Encoding and Primitive assemblies are unified, an implementation for PKCS#8 and SPKI keys can be provided in the abstract `AsymmetricAlgorithm` when importing PEM. Derived types will continue to override these members to provide support for additional PEM labels.

This is essentially addressing the feedback at https://github.com/dotnet/runtime/pull/34086#discussion_r404377337

> If we move the class around perhaps we'll remember to come fill this in.

So here we are.

### CODE REVIEW

**Comment by vcsjones** [comment]
I'm assuming I am going to need to make a docs PR separately to update the docs after this is merged, unless there is a tool to make that easier.

**Comment by vcsjones** [comment]
As a follow up, when `.Algorithms` is unified, I think we can remove some of the overrides since they can use what is here now (particularly, the encrypted PEMs and DSA only supports SPKI / PKCS8). But, I am unsure if that qualifies as a breaking change or not.

**Comment by bartonjs** [comment]
> As a follow up, I think we can remove some of the overrides in .Algorithms since they can use what is here now

In this change it should be fine to at least replace the body of the DSA methods with a call to base, no?   Assuming I'm a reliable authority here, removing the override entirely should actually be safe.  (The runtime handles the "override moved toward base" case gracefully.)

**Comment by bartonjs** [comment]
> I'm assuming I am going to need to make a docs PR separately to update the docs after this is merged, unless there is a tool to make that easier.

@carlossanlop What's the correct/easiest flow here?

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AsymmetricAlgorithm.cs [hunk 1]
```diff
@@ -165,55 +166,217 @@ protected virtual void Dispose(bool disposing)
...
+        ///   </para>
+        ///   <para>
+        ///   Types that override this method may support additional PEM labels.
+        ///   </para>
         /// </remarks>
-        public virtual void ImportFromPem(ReadOnlySpan<char> input) =>
-            throw new NotImplementedException(SR.NotSupported_SubclassOverride);
+        public virtual void ImportFromPem(ReadOnlySpan<char> input)
+        {
+            PemKeyImportHelpers.ImportPem(input, label => {
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AsymmetricAlgorithm.cs:365]:
```suggestion
            PemKeyImportHelpers.ImportPem(input, label =>
            {
```

**Comment by bartonjs** [comment]
Looks good, modulo one bad brace/newline ordering.  It'd be nice to see the DSA versions of these methods be removed to validate the usefulness of having this on the base class, but if something wonky stops that from being in the same change I'll sigh and move on.

**Comment by carlossanlop** [comment]
> > I'm assuming I am going to need to make a docs PR separately to update the docs after this is merged, unless there is a tool to make that easier.
> 
> What's the correct/easiest flow here?

@vcjones I see you're adding new exceptions, but you're also editing documentation that already existed.

We have a tool called DocsPortingTool that can help you port triple slash documentation to the dotnet-api-docs repo. The only problem is that it only works for _new_ documentation. So if you run the tool against the changes in this PR, only the new exceptions would get ported, but the existing summaries and exceptions that you are modifying would not get their new documentation ported.

So my recommendation would be to:

1. Clone dotnet/dotnet-api-docs.
2. Find the xmls and the APIs that you modified in this PR:
    a) `ImportFromEncryptedPem can be found here:` [AsymmetricAlgorithm.xml#L853-L864](https://github.com/dotnet/dotnet-api-docs/blob/11242b9ba428e75dcebb3e166c462fdcf2449493/xml/System.Security.Cryptography/AsymmetricAlgorithm.xml#L853-L864) and [AsymmetricAlgorithm.xml#L892-L906](https://github.com/dotnet/dotnet-api-docs/blob/11242b9ba428e75dcebb3e166c462fdcf2449493/xml/System.Security.Cryptography/AsymmetricAlgorithm.xml#L892-L906)
    b) `ImportFromPem` is located here: [AsymmetricAlgorithm.xml#L930-L943](https://github.com/dotnet/dotnet-api-docs/blob/11242b9ba428e75dcebb3e166c462fdcf2449493/xml/System.Security.Cryptography/AsymmetricAlgorithm.xml#L930-L943)
3. Change their descriptions of all the summaries, parameters and remarks to the string `To be added.` and also delete all the existing `<exception>` elements. 
4. Now clone the [DocsPortingTool](https://github.com/carlossanlop/DocsPortingTool) repo, build the project, and run this command:

```
DocsPortingTool \
        -Direction ToDocs \
        -Docs %REPOS_ROOT%\dotnet-api-docs\xml \
        -IntelliSense %REPOS_ROOT%\runtime\artifacts \
        -IncludedAssemblies System.Security.Cryptography \
        -IncludedTypes AsymmetricAlgorithm \
        -Save true
```
_// Replace %REPOS_ROOT% with the appropriate folder._

This will ensure all the new and modified docs get ported. Then you can submit a PR for the dotnet-api-docs repo.

Let me know if you have any questions.

**Comment by carlossanlop** [comment]
On a second thought, I just noticed you're not adding too many comments. It might be faster for you to simply copy your new triple slash docs and paste them in the locations I pointed out above.

**Review by bartonjs — APPROVED** [review]

**Comment by vcsjones** [comment]
@bartonjs 
> In this change it should be fine to at least replace the body of the DSA methods with a call to base, no? 

Oh, yeah that at least is I think is doable. I pushed those up where possible.

> Assuming I'm a reliable authority here, removing the override entirely should actually be safe. 

Yeah I can only come up with fairly contrived examples… like breaking reflection that uses `BindingFlags.DeclaredOnly`. I can remove the overrides entirely if you feel that is the better thing to do.

**Pull request merged**